services/media/service/migrations/committed/000022-transactional-outbox-inbox.sql

```diff
@@ -0,0 +1,235 @@
+GRANT SELECT, INSERT, UPDATE, DELETE ON app_hidden.outbox TO :DB_OWNER;
+
+-- Create the function to get the next batch of messages from the outbox table.
+DROP FUNCTION IF EXISTS app_hidden.next_outbox_messages(integer, integer);
+CREATE OR REPLACE FUNCTION app_hidden.next_outbox_messages(
+  max_size integer, lock_ms integer)
+    RETURNS SETOF app_hidden.outbox 
+    LANGUAGE 'plpgsql'
+
+AS $BODY$
+DECLARE 
+  loop_row app_hidden.outbox%ROWTYPE;
+  message_row app_hidden.outbox%ROWTYPE;
+  ids uuid[] := '{}';
+BEGIN
+
+  IF max_size < 1 THEN
+    RAISE EXCEPTION 'The max_size for the next messages batch must be at least one.' using errcode = 'MAXNR';
+  END IF;
+
+  -- get (only) the oldest message of every segment but only return it if it is not locked
+  FOR loop_row IN
+    SELECT * FROM app_hidden.outbox m WHERE m.id in (SELECT DISTINCT ON (segment) id
+      FROM app_hidden.outbox
+      WHERE processed_at IS NULL AND abandoned_at IS NULL
+      ORDER BY segment, created_at) order by created_at
+  LOOP
+    BEGIN
+      EXIT WHEN cardinality(ids) >= max_size;
+    
+      SELECT id, locked_until
+        INTO message_row
+        FROM app_hidden.outbox
+        WHERE id = loop_row.id
+        FOR NO KEY UPDATE NOWAIT;
+      
+      IF message_row.locked_until > NOW() THEN
+        CONTINUE;
+      END IF;
+      
+      ids := array_append(ids, message_row.id);
+    EXCEPTION WHEN lock_not_available THEN
+      CONTINUE;
+    END;
+  END LOOP;
+  
+  -- if max_size not reached: get the oldest parallelizable message independent of segment
+  IF cardinality(ids) < max_size THEN
+    FOR loop_row IN
+      SELECT * FROM app_hidden.outbox
+        WHERE concurrency = 'parallel' AND processed_at IS NULL AND abandoned_at IS NULL AND locked_until < NOW() 
+          AND id NOT IN (SELECT UNNEST(ids))
+        order by created_at
+    LOOP
+      BEGIN
+        EXIT WHEN cardinality(ids) >= max_size;
+
+        SELECT *
+          INTO message_row
+          FROM app_hidden.outbox
+          WHERE id = loop_row.id
+          FOR NO KEY UPDATE NOWAIT;
+
+        ids := array_append(ids, message_row.id);
+      EXCEPTION WHEN lock_not_available THEN
+        CONTINUE;
+      END;
+    END LOOP;
+  END IF;
+  
+  -- set a short lock value so the the workers can each process a message
+  IF cardinality(ids) > 0 THEN
+
+    RETURN QUERY 
+      UPDATE app_hidden.outbox
+        SET locked_until = clock_timestamp() + (lock_ms || ' milliseconds')::INTERVAL, started_attempts = started_attempts + 1
+        WHERE ID = ANY(ids)
+        RETURNING *;
+
+  END IF;
+END;
+$BODY$;
+
+CREATE INDEX outbox_segment_idx ON app_hidden.outbox (segment);
+CREATE INDEX outbox_created_at_idx ON app_hidden.outbox (created_at);
+CREATE INDEX outbox_processed_at_idx ON app_hidden.outbox (processed_at);
+CREATE INDEX outbox_abandoned_at_idx ON app_hidden.outbox (abandoned_at);
```

Comment:
Do we need to also add a index for `locked_until`, since its used for filtering
messages with parallel concurency?

Reply:
Good point - added it and switched to the idempotent `ax_define` variant.